New User Registration

YanaDePauw opened the pull request:
## Description
Addition of the registration and create profile components.

Depends on REST API PR https://github.com/DSpace/DSpace/pull/2763

## Instructions for Reviewers
This PR adds support in the form of service methods and components to support self-registration and creation of a profile for a new user.

List of changes in this PR:
* Component to display a "Register" page
* Component for the follow up "Create Profile" page
* New service to interact with the registration endpoint
* New method in the existing eperson service to create a profile based on a token

To verify that the functionality works, a new registration can be completed, the following can be done:
* Click on the login button
* Click on the "New user? Click here to register" link
* Fill in a valid email address and click register
* An email should have been sent to the provided email address with a link to the create profile page.
* Navigate to the url provided in the email, and fill out profile details
* Submit the form and verify that you are logged in as the new user.

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

atarix83 commented:
@YanaDePauw
I can't try this entirely, because even if I run over [related REST PR](https://github.com/DSpace/DSpace/pull/2763), I have no email server locally to test user registration from url provided in email.
I think is better to wait until REST PR is merged to test over official demo.

Meanwhile I suggest to remove every reference to the user registration within the auth module, because at the beginning registration was supposed to be implemented with the auth module. Some examples above:

- https://github.com/atmire/dspace-angular/blob/8fc2309d488c21935f9b3595e68dfb957e8eb594/src/app/core/auth/auth.actions.ts#L34-L36
- https://github.com/atmire/dspace-angular/blob/8fc2309d488c21935f9b3595e68dfb957e8eb594/src/app/core/auth/auth.service.ts#L277
- https://github.com/atmire/dspace-angular/blob/8fc2309d488c21935f9b3595e68dfb957e8eb594/src/app/core/auth/auth.effects.ts#L142

YanaDePauw commented:
@atarix83 
I removed the non-used registration references that i could find.

@artlowel 
I fixed the following things: 
- register-email.title
- "If press return or enter on the new user registration or create profile forms" 
→ I added enter support on on both forms but for some reason it doesn't trigger in the create profile component. 
- "You can't submit the create profile form until you click outside the password box. It would be better to check the disabled status onKeyUp rather than onBlur"
→ Changed this to be on debounce
